stm32h7 ethernet driver support

KozhinovAlexander opened the pull request:
Currently I have first working driver prototype under [my corresponding development branch](https://github.com/Nukersson/zephyr/tree/stm32h7xx_eth_drv_support). This driver were successfully tested on following Zephyr samples:

1. [Socket Dumb HTTP Server](https://docs.zephyrproject.org/latest/samples/net/sockets/dumb_http_server/README.html)
2. [Socket Multithreaded Dumb HTTP Server](https://docs.zephyrproject.org/latest/samples/net/sockets/dumb_http_server_mt/README.html)
3. [Socket Echo Server](https://docs.zephyrproject.org/latest/samples/net/sockets/echo/README.html#socket-echo-server)
4. [Asynchronous Socket Echo Server](https://docs.zephyrproject.org/latest/samples/net/sockets/echo_async/README.html)
5. [Socket HTTP GET Example](https://docs.zephyrproject.org/latest/samples/net/sockets/http_get/README.html#sockets-http-get)
6. [Socket Big HTTP Download Example](https://docs.zephyrproject.org/latest/samples/net/sockets/big_http_download/README.html)

erwango commented:
^^ @i-and, @KwonTae-young, @gmarull

gmarull commented on boards/arm/nucleo_h745zi_q/Kconfig.defconfig:
```diff
@@ -43,4 +43,24 @@ choice STM32H7_DUAL_CORE_BOOT
+	# proper ETH_DMA transitions
+	default y
+
+config USE_STM32_HAL_RCC_EX
```
Does board code need to access RCC? Sounds like this is not in the right place

Nukersson replied:
Yes, it needed to access RCC_EX (see [RCC_EX support issue](https://github.com/zephyrproject-rtos/hal_stm32/pull/56) in **hal_stm32** thread).
@gmarull What would be your proposal?

erwango replied:
@Nukersson, should be done in drivers/ethernet/Kconfig.stm32_hal:
```
menuconfig ETH_STM32_HAL
	bool "STM32 HAL Ethernet driver"
	select USE_STM32_HAL_ETH
+     select USE_STM32_HAL_RCC_EX if ....
	help
           ...
```

Nukersson replied:
@erwango Thank you for very helpful advice. The corresponding definition were moved to suggested KConfig file.

gmarull commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -36,8 +36,9 @@ config ETH_STM32_HAL_RX_THREAD_PRIO
-	default y
-	depends on SOC_SERIES_STM32F7X
+	default y if SOC_SERIES_STM32F7X
+	default n if SOC_SERIES_STM32H7X
```
is this required?

Nukersson replied:
Yes, it is. The **Eth. MAC - AHB** is not able to access **DTCM RAM** (see **Table 2** on **p.108** of [reference manual of H745](https://www.st.com/resource/en/reference_manual/dm00176879-stm32h745755-and-stm32h747757-advanced-armbased-32bit-mcus-stmicroelectronics.pdf)). Therefore it must be disabled for H7 series.

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,18 +32,112 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#define PHY_BSR     ((uint16_t)0x0001U)
+#endif
+
+#ifndef PHY_LINKED_STATUS
```
Why checking `ifndef`?

Nukersson replied:
> Why checking `ifndef`?

In case of PHY_BSR defined somewhere else. Should `#ifndef` be deleted?

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,18 +32,112 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define GET_TX_DESC_LIST_PTR(ps_heth)     (&(((ETH_HandleTypeDef*)ps_heth)->TxDescList))
+
+#define GET_FIRST_DMA_TX_DESC(ps_heth)                ((ETH_DMADescTypeDef*)(((ETH_HandleTypeDef*)ps_heth)->Init.TxDesc))
```
Too long lines here and below. Could you go through the whole eth driver file and try to shorten them to 80 chars.

Nukersson replied:
> Too long lines here and below. Could you go through the whole eth driver file and try to shorten them to 80 chars.

done

Nukersson replied:
done

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,18 +32,112 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #error DTCM for DMA buffer is activated but zephyr,dtcm is not present in dts
 #endif
 
+static HAL_StatusTypeDef ETH_RD_PHY_REG( ETH_HandleTypeDef* ps_heth,
```
Function name does not follow convention.

Nukersson replied:
> Function name does not follow convention.

renamed to `read_eth_phy_register`

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,18 +32,112 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+static HAL_StatusTypeDef ETH_RD_PHY_REG( ETH_HandleTypeDef* ps_heth,
+                                         uint32_t u32_PHYAddr,
+                                         uint32_t u32_PHYReg,
+                                         uint32_t* pu32_RegVal) {
```
function starting { should be in separate line

Nukersson replied:
> function starting { should be in separate line

done

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -60,6 +154,7 @@ static void set_mac_to_translator_addr(uint8_t *mac_addr)
 static void enable_canbus_eth_translator_filter(ETH_HandleTypeDef *heth,
 						uint8_t *mac_addr)
 {
+	// @TODO: must be done for STM32H7 series also!
```
no // comments

Nukersson replied:
> no // comments

done

Nukersson replied:
done

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -72,6 +167,15 @@ static inline void disable_mcast_filter(ETH_HandleTypeDef *heth)
+	MACFilterConf.PassAllMulticast = ENABLE;
+	MACFilterConf.HachOrPerfectFilter = DISABLE;
+	HAL_ETH_GetMACFilterConfig(heth, &MACFilterConf);
+	k_sleep(K_MSEC(1));
```
Why is sleep needed?

Nukersson replied:
> Why is sleep needed?

To hold same runtime of the function `disable_mcast_filter` as t ws before. Should it be deleted?

jukkar replied:
Using k_sleep() might cause a context switch to other thread, would that be an issue?
Other alternative is to use `k_busy_wait()` here. The original code is calling `k_sleep()` too so perhaps this is not an issue.

Nukersson replied:
@jukkar Since the function `HAL_ETH_SetMACFilterConfig()` was already executed at this time point, the context switch would not be a problem.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -72,6 +167,15 @@ static inline void disable_mcast_filter(ETH_HandleTypeDef *heth)
 	__ASSERT_NO_MSG(heth != NULL);
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	ETH_MACFilterConfigTypeDef MACFilterConf;
```
This addition looks very crowded, some more empty lines would do good.

Nukersson replied:
> This addition looks very crowded, some more empty lines would do good.

done

Nukersson replied:
added empty lines

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -98,43 +203,78 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 
-	total_len = net_pkt_get_len(pkt);
-	if (total_len > ETH_TX_BUF_SIZE) {
+	u32_total_len = net_pkt_get_len(pkt);
```
"hungarian notation" ?

Nukersson replied:
> "hungarian notation" ?

Yes, it is. Changed back to `total_len`.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -98,43 +203,78 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	uint8_t *dma_buffer;
 	int res;
-	uint16_t total_len;
+	uint32_t u32_total_len;
```
What is the purpose of this change? If you are worried about the size of the var, just change the type of total_len to size_t.

Nukersson replied:
> What is the purpose of this change? If you are worried about the size of the var, just change the type of total_len to size_t.

changed to `size_t`

Nukersson replied:
changed to `size_t`

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -177,11 +325,13 @@ static struct net_pkt *eth_rx(struct device *dev, uint16_t *vlan_tag)
 	ETH_HandleTypeDef *heth;
-	__IO ETH_DMADescTypeDef *dma_rx_desc;
+#if !defined(CONFIG_SOC_SERIES_STM32H7X)
+    __IO ETH_DMADescTypeDef *dma_rx_desc;
```
indentation is wrong here

Nukersson replied:
> indentation is wrong here

changed to correct one.

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -98,43 +203,78 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+#endif
+
+	dma_tx_desc = GET_FIRST_DMA_TX_DESC(heth);
+    // dma_tx_desc = ((ETH_DMADescTypeDef*)(heth->TxDescList.TxDesc[u32_cur_tx_desc]));
```
Commented code should be deleted

Nukersson replied:
> Commented code should be deleted

done

gmarull reviewed (dismissed):
I think there are a few style/code issues that should be fixed first.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -305,8 +489,9 @@ static void rx_thread(void *arg1, void *unused1, void *unused2)
 			/* semaphore timeout period expired, check link status */
-			if (HAL_ETH_ReadPHYRegister(&dev_data->heth, PHY_BSR,
-				(uint32_t *) &status) == HAL_OK) {
+            hal_ret = ETH_RD_PHY_REG(&dev_data->heth, PHY_ADDR, PHY_BSR,
```
indentation

Nukersson replied:
changed to correct one

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -401,7 +586,7 @@ static int eth_initialize(struct device *dev)
 	}
 
-	heth = &dev_data->heth;
+	heth = &(dev_data->heth);
```
unnecessary change, can be removed

Nukersson replied:
removed braces

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -412,17 +597,33 @@ static int eth_initialize(struct device *dev)
 	}
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+    // Tx config init:
```
indentation

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -436,19 +637,50 @@ static int eth_initialize(struct device *dev)
 			0, K_NO_WAIT);
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+    for(uint32_t idx = 0; idx < ETH_RX_DESC_CNT; idx ++) {
```
indentation
also you should have a space between "for" and (

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -436,19 +637,50 @@ static int eth_initialize(struct device *dev)
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+    for(uint32_t idx = 0; idx < ETH_RX_DESC_CNT; idx ++) {
+        hal_ret = HAL_ETH_DescAssignMemory(heth, idx, dma_rx_buffer[idx], NULL);
+        if(hal_ret != HAL_OK) {
```
space between if and (

Nukersson replied:
done

jukkar reviewed (commented):
some mostly cosmetic nits

Nukersson commented:
@erwango Can you please test the driver with `samples/net/sockets/dumb_http_server_mt` or `samples/net/sockets/dumb_http_server` sample project and one of the stm32h7 ( h745, h747 ) boards?
Please write also feedback here.

I've currently tested it with my `nucleo-h745zi-q` and it works fine.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+																TxDescList))
+
+#define GET_FIRST_DMA_TX_DESC(heth)  ((ETH_DMADescTypeDef *) \
+									  (((ETH_HandleTypeDef *)heth)-> \
```
the last \ seems to be redundant as the next line is empty
indentation is also way off

Nukersson replied:
corrected

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+									  (((ETH_HandleTypeDef *)heth)-> \
+																Init.TxDesc))
+#define GET_FIRST_DMA_RX_DESC(heth)  ((ETH_DMADescTypeDef *) \
+									  (((ETH_HandleTypeDef *)heth)-> \
```
same here

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+																Init.RxDesc))
+
+#define IS_ETH_DMATXDESC_OWN(dma_tx_desc)   (((ETH_DMADescTypeDef *) \
+												dma_tx_desc)->DESC3 & \
```
and here

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+												dma_tx_desc)->DESC3 & \
+												ETH_DMATXNDESCRF_OWN)
+#define GET_DMA_TX_BUFFER1_ADDR(dma_tx_desc) (((ETH_DMADescTypeDef *) \
+												dma_tx_desc)->DESC0)
```
and here

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#define ETH_DMA_TRANSMIT_TIMEOUT        20U
+
+static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
+											   uint32_t PHYAddr,
```
can you check the indentation here and below, there are indent issues in so many places that I do not bother to comment them all

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -98,15 +227,16 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	__ASSERT_NO_MSG(dev_data != NULL);
 
-	heth = &dev_data->heth;
+	heth = &(dev_data->heth);
```
the original code is just fine, please remove () here

Nukersson replied:
done

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -291,7 +498,7 @@ static void rx_thread(void *arg1, void *unused1, void *unused2)
 				dev_data->link_up = true;
 				net_eth_carrier_on(get_iface(dev_data,
-							     vlan_tag));
+								   vlan_tag));
```
original code indentation was correct, please remove the change

Nukersson replied:
done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
 #if defined(CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER) && \
-	    !DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_dtcm), okay)
+			!DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_dtcm), okay)
```
Please avoid changing code when this does not relate directly to your change. This make change harder to review.
If clean up is required, please do it in a separate commit.
Though, in this case, indentation was correct

Nukersson replied:
Ok. I understand this way. Will be changed with next commit.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#if defined(CONFIG_NOCACHE_MEMORY)
+#define NO_CACHE_DATA __nocache
 #else
-static ETH_DMADescTypeDef dma_rx_desc_tab[ETH_RXBUFNB] __aligned(4);
-static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __aligned(4);
-static uint8_t dma_rx_buffer[ETH_RXBUFNB][ETH_RX_BUF_SIZE] __aligned(4);
-static uint8_t dma_tx_buffer[ETH_TXBUFNB][ETH_TX_BUF_SIZE] __aligned(4);
-#endif /* CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER */
```
When making changes that apply to other series, this should be explained and put in dedicated commit.
We need to ensure the existing code is not broken or understand there are good reasons to change it.
In any case, don't modify it silently.

Nukersson replied:
Comes with next push.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#error "Error: CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER is not applicabble \
+		for STM32H7 series in this module!"
+#endif
```
It would be nice to format this error message in a different way.
Current proposal brokes github syntax highligher, which makes the review of the remaining code harder.

Nukersson replied:
done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#endif
+
+#else
+
```
Don't abuse of empty lines

Nukersson replied:
deleted empty lines

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#endif
+
+static ETH_DMADescTypeDef dma_rx_desc_tab[ETH_RXBUFNB] NO_CACHE_DATA \
+														ETH_DMA_MEM_SECTION;
```
Please fix indentation

Nukersson replied:
done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -60,18 +175,31 @@ static void set_mac_to_translator_addr(uint8_t *mac_addr)
 	heth->Instance->MACA1LR = (mac_addr[3] << 24U) | (mac_addr[2] << 16U) |
 				  (mac_addr[1] << 8U) | mac_addr[0];
-	/*enable filter 1 and ignore byte 5 and 6 for filtering*/
+	/* enable filter 1 and ignore byte 5 and 6 for filtering */
```
Please perform clean up in a dedicated commit

Nukersson replied:
Will be leaved as in master currently

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -60,18 +175,31 @@ static void set_mac_to_translator_addr(uint8_t *mac_addr)
 				  ETH_MACA1HR_MBC_HBits7_0;
 }
-#endif /*CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR*/
+#endif /* CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR */
```
Please perform clean up in a dedicated commit

Nukersson replied:
done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -60,18 +175,31 @@ static void set_mac_to_translator_addr(uint8_t *mac_addr)
+	HAL_ETH_SetMACFilterConfig(heth, &MACFilterConf);
+
+	k_sleep(K_MSEC(1));
+#else  /* CONFIG_SOC_SERIES_STM32H7X) */
```
remove the extra ')'

Nukersson replied:
done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -90,6 +218,7 @@ static inline void disable_mcast_filter(ETH_HandleTypeDef *heth)
 	tmp = heth->Instance->MACFFR;
 	k_sleep(K_MSEC(1));
 	heth->Instance->MACFFR = tmp;
+#endif  /* CONFIG_SOC_SERIES_STM32H7X) */
```
remove the extra ')'

Nukersson replied:
done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -98,15 +227,16 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	uint8_t *dma_buffer;
 	int res;
-	uint16_t total_len;
+	size_t total_len;
```
Please fix this in a dedicated commit

Nukersson replied:
moved to dedicated commit

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -117,24 +247,57 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+	const uint32_t u32_cur_tx_desc = 0;  /* heth->TxDescList.CurTxDesc; */
+#endif
+
+	dma_tx_desc = GET_FIRST_DMA_TX_DESC(heth);
+	while (IS_ETH_DMATXDESC_OWN(dma_tx_desc) != (uint32_t)RESET) {
```
Please move existing code modification to dedicated commits

Nukersson replied:
Comes with next push.

erwango reviewed (changes requested):
@Nukersson, Txs again for this work. Please find some comments:

One of the first things we check is that the existing code is not broken.
It could be modified, but this requires explanations so ideally it should be in dedicated commits.
Then, in the commit that add H7 support, ideally there should be no 'red' (line deletions) and all 'green' should be under "#if defined(CONFIG_SOC_SERIES_STM32H7X)...

Also please take car to indentation (1tab = 8 spaces)

Nukersson commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -36,8 +36,9 @@ config ETH_STM32_HAL_RX_THREAD_PRIO
-	depends on SOC_SERIES_STM32F7X
+	default y if SOC_SERIES_STM32F7X
+	default n if SOC_SERIES_STM32H7X
+	depends on SOC_SERIES_STM32F7X && SOC_SERIES_STM32H7X
```
Yes. You're right! Will be corrected.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -28,22 +28,137 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #endif
 
+static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef * heth,
+											   uint32_t PHYAddr,
```
> Hmm, is you tab size other than 8 chars, the parameter indentation is way off.

Currently my TAB size is 4 characters. In the line 36 here 11 TABs and 3 spaces are placed. It looks like the editor of GitHub displays it different than my VisualStudio Code editor.

Nukersson replied:
> Hmm, is you tab size other than 8 chars, the parameter indentation is way off.

TAB size is now 8 characters. All the identitatiot errors were fixed s.t. `checkpatch.pl` runs without warnings on my commits.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -146,6 +309,14 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_Transmit: failed: %d, %d, %d, %d",
+				hal_ret, heth->ErrorCode, heth->DMAErrorCode, heth->gState);
```
We're totally right. I will change all my introduced `LOG_ERR()` messages to simple `<Function Name>: failed!` string. Thus the pointed kind of errors can be always avoided.

Commit:
drivers: pinumx: stm32: pinmux_stm32h7: added ethernet pins

added ethernet pins definition for stm32h7 SoC series

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

Commit:
drivers: ethernet: stm32: pre-add stm32h7 support

added changes that apply to other than stm32h7 series
and prepares stm32h7 series integration:

eth_tx/eth_rx: changed type of total_len
added necessary defines for
    tx_desc and own flag extractions
added read_eth_phy_register() function
    and corresponding defines
added ETH_DMA_MEM and CACHE defines
following memory buffers declarations moved out of
    its ifdef preprocessor macro:
        dma_rx_desc_tab, dma_tx_desc_tab,
        dma_rx_buffer, dma_tx_buffer
added hal_ret stm32 HAL status checks

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

Commit:
boards: nucleo_h745zi_q: added ethernet support

KConfig.defconfig: enabled NETWORKING support
pinmux.c: added ethernet pins definition for this board
m7 yaml file becomes netif:eth support enabled

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

Commit:
drivers: ethernet: stm32: add stm32h7 support

add DTCM caching management in KConfig for stm32h7
add NOCACHE_MEMORY support for stm32h7 M7 CPU series
add HAL_RCC_EX support for stm32h7 series
implemented stm32h7 support within stm32 eth driver

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -94,3 +97,10 @@ config ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS
 endif # ETH_STM32_HAL
+
+config NOCACHE_MEMORY
+	default y if SOC_SERIES_STM32H7X
```
`if SOC_SERIES_STM32H7X` is superfluous in this case.

Nukersson replied:
> `if SOC_SERIES_STM32H7X` is superfluous in this case.

But then it will be enabled for F4/F7 series also. Would it be okay?

erwango replied:
Since you mention `depends on SOC_SERIES_STM32H7X` just below, I think we're safe

Nukersson replied:
Will be done within next push.

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -94,3 +97,10 @@ config ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS
+	depends on SOC_SERIES_STM32H7X
+	help
+	  ETH DMA memory cache issues are not optimal for STM32H7 series.
+	  Thererfore no-cache memory should be enabled.
```
'Therefore'

Nukersson replied:
Will be changed with next push. 👯

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -36,8 +38,9 @@ config ETH_STM32_HAL_RX_THREAD_PRIO
 
 config ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER
 	bool "Use DTCM for DMA buffers"
-	default y
-	depends on SOC_SERIES_STM32F7X
+	default y if SOC_SERIES_STM32F7X
+	default n if SOC_SERIES_STM32H7X
+	depends on SOC_SERIES_STM32F7X || SOC_SERIES_STM32H7X
```
If default n for H7, why not leaving this as is ("depends on F7")?

Nukersson replied:
You're right. It was just remnant from dev stage.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -68,10 +136,41 @@ static void enable_canbus_eth_translator_filter(ETH_HandleTypeDef *heth,
 }
 #endif /*CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR*/
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
+						uint32_t PHYAddr,
+						uint32_t PHYReg,
+						uint32_t *RegVal)
+{
+	return HAL_ETH_ReadPHYRegister(heth, PHYAddr, PHYReg, RegVal);
+}
+#else
+static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
+						uint32_t PHYAddr,
+						uint32_t PHYReg,
+						uint32_t *RegVal)
+{
+	ARG_UNUSED(PHYAddr);
+	return HAL_ETH_ReadPHYRegister(heth, PHYReg, RegVal);
+}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Putting `#ifdef` inside function would enhance readability.

Nukersson replied:
Should then all `#if defined(CONFIG_SOC_SERIES_STM32H7X)` be replaced by `#ifdef CONFIG_SOC_SERIES_STM32H7X` ?

erwango replied:
Sorry, I was not clear enough, you should be able to do something like:
```
static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
						uint32_t PHYAddr,
						uint32_t PHYReg,
						uint32_t *RegVal)
{
#if defined(CONFIG_SOC_SERIES_STM32H7X)
	return HAL_ETH_ReadPHYRegister(heth, PHYAddr, PHYReg, RegVal);
#else
	ARG_UNUSED(PHYAddr);
	return HAL_ETH_ReadPHYRegister(heth, PHYReg, RegVal);
#endif
}
```

Nukersson replied:
Will be done within next push.

erwango reviewed (changes requested):
Few more suggestions

Nukersson commented:
Hello all. I've done following things as it was suggested by reviewers:

1. Changed tab size to 8 spaces and corrected line lengths.
2. Splitter commits into following parts:
    - Pins definitions: `drivers: pinumx: stm32: pinmux_stm32h7: added ethernet pins` @ 45c17fc 
    - stm32h7 series independent changes (should work as in current upstream/master): `drivers: ethernet: stm32: pre-add stm32h7 support` @ 
9972b09
    - adding support of `stm32h7 series` @ 6ce77a0
    - adding support of `nucleo_h745zi_q_m7` board @ 5f18f61

Pleas feel free for new review and thank you!

cfriedt commented:
Last time I looked at an stm ethernet driver, it became apparent that it was just a designware block. Is this ethernet block different from e.g. the stm32f479?
https://github.com/torvalds/linux/blob/master/drivers/net/ethernet/stmicro/stmmac/dwmac-stm32.c

Nukersson commented:
> Last time I looked at an stm ethernet driver, it became apparent that it was just a designware block. Is this ethernet block different from e.g. the stm32f479?
> https://github.com/torvalds/linux/blob/master/drivers/net/ethernet/stmicro/stmmac/dwmac-stm32.c

Thank you for your notice. 

1. According to [this](https://www.google.com/url?q=https://www.st.com/resource/en/application_note/dm00337702-migration-of-microcontroller-applications-from-stm32f7-series-to-stm32h7x3-line-microcontrollers-stmicroelectronics.pdf&sa=U&ved=2ahUKEwiUoKzdkqXqAhWt5KYKHbVtBMMQFjAAegQICxAB&usg=AOvVaw1sQn7KfjoE-P3Y2W5TnGLB) there are differences between stm32h7x3 and stm32f7. Probably @erwango can better answer your question.

2. Thank you for the link to stm32 driver implementation. For me it looks like a stm32mp1 ether driver layer focussing on linux support.

3. You're right. There are many already working stm32h7 Ethernet drivers. One of them I mentioned in #24767 issue and there should be one from Olimex also. Maybe for now there are even more.

4. But, we're focusing here on integration of stm32h7 Ethernet driver into zephyr rtos. Thus some effort needed, even if already working implementation for lwip for example would be taken.


If you're interested, you can take a look at following issue #24767. There you can find the discussion concerning stm32h7 series Ethernet driver implementation for Zephyr.

erwango commented:
> According to this there are differences between stm32h7x3 and stm32f7 for example. Probably @erwango can better answer your question.

There are 4 variants of ethernet IP accross STM32 series. H7 one is indeed different from F4 one

ABOSTM reviewed (changes requested):
Unfortunately there is a regression on nucleo_f767zi:
samples/net/sockets/dumb_http_server doesn't work anymore.
Web server is no more reachable.
Regression is introduced by 2nd commit : 'drivers: ethernet: stm32: pre-add stm32h7 support'

Nukersson commented:
> Unfortunately there is a regression on nucleo_f767zi:
> samples/net/sockets/dumb_http_server doesn't work anymore.
> Web server is no more reachable.
> Regression is introduced by 2nd commit : 'drivers: ethernet: stm32: pre-add stm32h7 support'
> 

Could you find, what is the problem? I can't test it on the board, since I do not have it. 
From my side, I will test it later on ma nucleo-h745zi-q

ABOSTM commented:
After more investigations, it appears this is not a regression, but rather an instability on nucleo_f767zi board, which is already present before your PR. I will create a new issue for that.

Nukersson commented:
> After more investigations, it appears this is not a regression, but rather an instability on nucleo_f767zi board, which is already present before your PR. I will create a new issue for that.

Thank you. You're saving my time! :)

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -44,15 +45,43 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
 #define ETH_DMA_MEM	__dtcm_noinit_section
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#error "Error: CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER is not applicable" \
+	" for STM32H7 series in this module!"
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
```
done

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -117,24 +207,67 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
-	dma_tx_desc = heth->TxDesc;
-	while ((dma_tx_desc->Status & ETH_DMATXDESC_OWN) != (uint32_t)RESET) {
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	const uint32_t u32_cur_tx_desc = 0;  /* heth->TxDescList.CurTxDesc; */
```
done

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -436,19 +624,50 @@ static int eth_initialize(struct device *dev)
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_Start{_IT} failed");
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
You're definitely right! Will be changed with next push.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -65,11 +94,19 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define NO_CACHE
 #endif
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && !defined(CONFIG_NOCACHE_MEMORY)
+#error "Error: CONFIG_NOCACHE_MEMORY must be defined for STM32H7 series!"
+#endif
+
```
Removed above decompiler error message.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -85,6 +122,7 @@ static void set_mac_to_translator_addr(uint8_t *mac_addr)
 static void enable_canbus_eth_translator_filter(ETH_HandleTypeDef *heth,
 						uint8_t *mac_addr)
 {
+	/* @TODO: must be done for STM32H7 series also! */
```
I've added the support for `enable_canbus_eth_translator_filter`. It is pretty straightforward and almost same as f7 series. But I can not test it properly. Thus this conversation should stay unresolved for a while or until somebody can test it properly.

ABOSTM commented:
I tested samples
* dumb_http_server 
* big_http_download

on boards:
* nucleo_f429zi
* nucleo_f767zi
* nucleo_f746zg
* nucleo_f207zg

All is passed with this PR.
(Even if some instability already present before this PR on nucleo_f767zi: #26519)

Nukersson commented:
> I tested samples
> 
> * dumb_http_server
> * big_http_download
> 
> on boards:
> 
> * nucleo_f429zi
> * nucleo_f767zi
> * nucleo_f746zg
> * nucleo_f207zg
> 
> All is passed with this PR.
> (Even if some instability already present before this PR on nucleo_f767zi: #26519)

Thank you very very much!

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -152,25 +199,65 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
+	 * Resume processing the Transmit descriptors.
+	 */
+	if ((heth->Instance->DMACSR & ETH_DMACSR_TBU) != (uint32_t)RESET) {
+		/* @TODO: Must be done or skipped! */
```
Rather than merging a commented section, I'd prefer this whole piece of code being left as a follow up PR with a tag "Block: HW Tests"

Nukersson replied:
Will be deleted within next push.

erwango commented:
@Nukersson we'll rely on your tests for the validation on H7 series. It seems there is no variant of eth IP within H7 series, so testing on other H7 target would be reqdundant.

Nukersson commented:
> @Nukersson we'll rely on your tests for the validation on H7 series. It seems there is no variant of eth IP within H7 series, so testing on other H7 target would be redundant.

Do you mean under "`no variant of eth IP within H7 series`" - the eth HW module in all H7 ist the same!

erwango commented:
> > @Nukersson we'll rely on your tests for the validation on H7 series. It seems there is no variant of eth IP within H7 series, so testing on other H7 target would be redundant.
> 
> Do you mean under "`no variant of eth IP within H7 series`" - the eth HW module in all H7 ist the same!

Exactly, so we'll rely on your tests and won't do more testing on H7.

Nukersson commented:
> I tested samples
> 
> * dumb_http_server
> * big_http_download
> 
> on boards:
> 
> * nucleo_f429zi
> * nucleo_f767zi
> * nucleo_f746zg
> * nucleo_f207zg
> 
> All is passed with this PR.
> (Even if some instability already present before this PR on nucleo_f767zi: #26519)

Have you used default 6.7MiB file with `big_http_download` example?

lochej commented:
I have tested the implementation proposed by @Nukersson on his branch to get the ETH driver for my H7 board.

I am still not very familiar with git pull requests but I still want to contribute and share an improvement.

It was tested on my **nucleo-H743ZI board**.

The **HAL_ETH_Transmit** function in **eth_stm32_hal.c** sometimes crashed with a **Timeout**. This breaks the net driver and make ping/ICMP and the board just stops working.

My suggestion is to switch to an interrupt based method and waiting for the **Transmit Complete interrupt** with a semaphore.

1. Add the new semaphore do **eth_stm32_hal_priv.h**:
```
/* Device run time data */
struct eth_stm32_hal_dev_data {
	struct net_if *iface;
	uint8_t mac_addr[6];
	ETH_HandleTypeDef heth;
	/* clock device */
	struct device *clock;
	struct k_mutex tx_mutex;
	struct k_sem tx_int_sem; //<-- New  ethTX complete sempaphore to add
	struct k_sem rx_int_sem;
	K_THREAD_STACK_MEMBER(rx_thread_stack,
		CONFIG_ETH_STM32_HAL_RX_THREAD_STACK_SIZE);
	struct k_thread rx_thread;
	bool link_up;
};
```

2. Init the **tx_int_sem** correctly in **eth_stm32_hal.c** funcion **eth_init**:
```
static int eth_initialize(struct device *dev)
{
       ....

	/* Initialize semaphores */
	k_mutex_init(&dev_data->tx_mutex);
	k_sem_init(&dev_data->rx_int_sem, 0, UINT_MAX);
	k_sem_init(&dev_data->tx_int_sem,0,UINT_MAX);

        ....
}
```


3. Add the TX complete callback from the HAL in **eth_stm32_hal.c**:
```
void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef * heth_handle)
{
	__ASSERT_NO_MSG(heth_handle != NULL);

	struct eth_stm32_hal_dev_data *dev_data =
			CONTAINER_OF(heth_handle, struct eth_stm32_hal_dev_data, heth);

	__ASSERT_NO_MSG(dev_data != NULL);

	k_sem_give(&dev_data->tx_int_sem);
}
```

4. Change the **eth_tx** function to use **HAL_ETH_TransmitIT** and new semaphore in **eth_stm32_hal.c**:
```
static int eth_tx(struct device *dev, struct net_pkt *pkt)
{
	struct eth_stm32_hal_dev_data *dev_data = DEV_DATA(dev);
	ETH_HandleTypeDef *heth;
	uint8_t *dma_buffer;
	int res;
	size_t total_len;
	__IO ETH_DMADescTypeDef *dma_tx_desc;
	HAL_StatusTypeDef hal_ret = HAL_OK;

	__ASSERT_NO_MSG(pkt != NULL);
	__ASSERT_NO_MSG(pkt->frags != NULL);
	__ASSERT_NO_MSG(dev != NULL);
	__ASSERT_NO_MSG(dev_data != NULL);

	heth = &dev_data->heth;

	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);

	total_len = net_pkt_get_len(pkt);
	if (total_len > ETH_TX_BUF_SIZE) {
		LOG_ERR("PKT to big");
		res = -EIO;
		goto error;
	}

#if defined(CONFIG_SOC_SERIES_STM32H7X)
	const uint32_t cur_tx_desc_idx = 0;  /* heth->TxDescList.CurTxDesc; */
#endif

	dma_tx_desc = GET_FIRST_DMA_TX_DESC(heth);
	while (IS_ETH_DMATXDESC_OWN(dma_tx_desc) != (uint32_t)RESET) {
		k_yield();
	}

#if defined(CONFIG_SOC_SERIES_STM32H7X)
	dma_buffer = dma_tx_buffer[cur_tx_desc_idx];
#else
	dma_buffer = (uint8_t *)(dma_tx_desc->Buffer1Addr);
#endif /* CONFIG_SOC_SERIES_STM32H7X */

	if (net_pkt_read(pkt, dma_buffer, total_len)) {
		res = -EIO;
		goto error;
	}

#if defined(CONFIG_SOC_SERIES_STM32H7X)
	ETH_BufferTypeDef tx_buffer_def[ETH_TXBUFNB];

	memset(tx_buffer_def, 0, ETH_TXBUFNB*sizeof(ETH_BufferTypeDef));

	tx_buffer_def[cur_tx_desc_idx].buffer = dma_buffer;
	tx_buffer_def[cur_tx_desc_idx].len = total_len;
	tx_buffer_def[cur_tx_desc_idx].next = NULL;

	tx_config.Length = total_len;
	tx_config.TxBuffer = tx_buffer_def;

	//hal_ret = HAL_ETH_Transmit(heth, &tx_config, ETH_DMA_TRANSMIT_TIMEOUT); //OLD BUGGY CODE sometime crashes with TIMEOUT
	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);

	if (hal_ret != HAL_OK) {
		LOG_ERR("HAL_ETH_TransmitIT {Frame}: failed! ret=%d err=%d",hal_ret,heth->ErrorCode);
		res = -EIO;
		goto error;
	}

	if(k_sem_take(&dev_data->tx_int_sem, K_MSEC(ETH_DMA_TRANSMIT_TIMEOUT))!=0)
	{
		LOG_ERR("HAL_ETH_TransmitIT timeout on tx_int_sem take");
		res = -EIO;
		goto error;
	}

#else

	hal_ret = HAL_ETH_TransmitFrame(heth, total_len);
#endif /* CONFIG_SOC_SERIES_STM32H7X */

	if (hal_ret != HAL_OK) {
		LOG_ERR("HAL_ETH_Transmit{Frame}: failed! ret=%d err=%d",hal_ret,heth->ErrorCode);
		res = -EIO;
		goto error;
	}

#if !defined(CONFIG_SOC_SERIES_STM32H7X)
	/* When Transmit Underflow flag is set, clear it and issue a
	 * Transmit Poll Demand to resume transmission.
	 */
	if ((heth->Instance->DMASR & ETH_DMASR_TUS) != (uint32_t)RESET) {
		/* Clear TUS ETHERNET DMA flag */
		heth->Instance->DMASR = ETH_DMASR_TUS;
		/* Resume DMA transmission*/
		heth->Instance->DMATPDR = 0;
		res = -EIO;
		goto error;
	}
#endif /* CONFIG_SOC_SERIES_STM32H7X */

	res = 0;
	error:
	k_mutex_unlock(&dev_data->tx_mutex);

	return res;
}
```

With these changes I have made the ethernet communication a lot faster and more reliable:
1. 0 packet losses for ICMP/ping with <2ms response time even with stress test 0.01s ping interval
   1. ping -i 0.01 192.0.2.1 
   2. 1038 packets transmitted
   3. 1038 received, 0% packet loss, time 10373ms
   4. rtt min/avg/max/mdev = 1.214/2.033/7.364/0.644 ms

2. net telnet sample works flawlessly with no connection break
3. net socket coap-server sample works perfectly
4. net socket dumb-http-server works perfectly


I hope you can integrate these into the driver meanwhile I learn how to use git and contribute correctly to the issues.

Nukersson commented:
> I have tested the implementation proposed by @Nukersson on his branch to get the ETH driver for my H7 board.
> 
> I am still not very familiar with git pull requests but I still want to contribute and share an improvement.
> 
> It was tested on my **nucleo-H743ZI board**.
> 
> The **HAL_ETH_Transmit** function in **eth_stm32_hal.c** sometimes crashed with a **Timeout**. This breaks the net driver and make ping/ICMP and the board just stops working.
> 
> My suggestion is to switch to an interrupt based method and waiting for the **Transmit Complete interrupt** with a semaphore.
> 
> 1. Add the new semaphore do **eth_stm32_hal_priv.h**:
> ```
> /* Device run time data */
> struct eth_stm32_hal_dev_data {
> 	struct net_if *iface;
> 	uint8_t mac_addr[6];
> 	ETH_HandleTypeDef heth;
> 	/* clock device */
> 	struct device *clock;
> 	struct k_mutex tx_mutex;
> 	struct k_sem tx_int_sem; //<-- New  ethTX complete sempaphore to add
> 	struct k_sem rx_int_sem;
> 	K_THREAD_STACK_MEMBER(rx_thread_stack,
> 		CONFIG_ETH_STM32_HAL_RX_THREAD_STACK_SIZE);
> 	struct k_thread rx_thread;
> 	bool link_up;
> };
> ```
> 
> 2. Init the **tx_int_sem** correctly in **eth_stm32_hal.c** funcion **eth_init**:
> ```
> static int eth_initialize(struct device *dev)
> {
>        ....
> 
> 	/* Initialize semaphores */
> 	k_mutex_init(&dev_data->tx_mutex);
> 	k_sem_init(&dev_data->rx_int_sem, 0, UINT_MAX);
> 	k_sem_init(&dev_data->tx_int_sem,0,UINT_MAX);
> 
>         ....
> }
> ```
> 
> 
> 3. Add the TX complete callback from the HAL in **eth_stm32_hal.c**:
> ```
> void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef * heth_handle)
> {
> 	__ASSERT_NO_MSG(heth_handle != NULL);
> 
> 	struct eth_stm32_hal_dev_data *dev_data =
> 			CONTAINER_OF(heth_handle, struct eth_stm32_hal_dev_data, heth);
> 
> 	__ASSERT_NO_MSG(dev_data != NULL);
> 
> 	k_sem_give(&dev_data->tx_int_sem);
> }
> ```
> 
> 4. Change the **eth_tx** function to use **HAL_ETH_TransmitIT** and new semaphore in **eth_stm32_hal.c**:
> ```
> static int eth_tx(struct device *dev, struct net_pkt *pkt)
> {
> 	struct eth_stm32_hal_dev_data *dev_data = DEV_DATA(dev);
> 	ETH_HandleTypeDef *heth;
> 	uint8_t *dma_buffer;
> 	int res;
> 	size_t total_len;
> 	__IO ETH_DMADescTypeDef *dma_tx_desc;
> 	HAL_StatusTypeDef hal_ret = HAL_OK;
> 
> 	__ASSERT_NO_MSG(pkt != NULL);
> 	__ASSERT_NO_MSG(pkt->frags != NULL);
> 	__ASSERT_NO_MSG(dev != NULL);
> 	__ASSERT_NO_MSG(dev_data != NULL);
> 
> 	heth = &dev_data->heth;
> 
> 	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
> 
> 	total_len = net_pkt_get_len(pkt);
> 	if (total_len > ETH_TX_BUF_SIZE) {
> 		LOG_ERR("PKT to big");
> 		res = -EIO;
> 		goto error;
> 	}
> 
> #if defined(CONFIG_SOC_SERIES_STM32H7X)
> 	const uint32_t cur_tx_desc_idx = 0;  /* heth->TxDescList.CurTxDesc; */
> #endif
> 
> 	dma_tx_desc = GET_FIRST_DMA_TX_DESC(heth);
> 	while (IS_ETH_DMATXDESC_OWN(dma_tx_desc) != (uint32_t)RESET) {
> 		k_yield();
> 	}
> 
> #if defined(CONFIG_SOC_SERIES_STM32H7X)
> 	dma_buffer = dma_tx_buffer[cur_tx_desc_idx];
> #else
> 	dma_buffer = (uint8_t *)(dma_tx_desc->Buffer1Addr);
> #endif /* CONFIG_SOC_SERIES_STM32H7X */
> 
> 	if (net_pkt_read(pkt, dma_buffer, total_len)) {
> 		res = -EIO;
> 		goto error;
> 	}
> 
> #if defined(CONFIG_SOC_SERIES_STM32H7X)
> 	ETH_BufferTypeDef tx_buffer_def[ETH_TXBUFNB];
> 
> 	memset(tx_buffer_def, 0, ETH_TXBUFNB*sizeof(ETH_BufferTypeDef));
> 
> 	tx_buffer_def[cur_tx_desc_idx].buffer = dma_buffer;
> 	tx_buffer_def[cur_tx_desc_idx].len = total_len;
> 	tx_buffer_def[cur_tx_desc_idx].next = NULL;
> 
> 	tx_config.Length = total_len;
> 	tx_config.TxBuffer = tx_buffer_def;
> 
> 	//hal_ret = HAL_ETH_Transmit(heth, &tx_config, ETH_DMA_TRANSMIT_TIMEOUT); //OLD BUGGY CODE sometime crashes with TIMEOUT
> 	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
> 
> 	if (hal_ret != HAL_OK) {
> 		LOG_ERR("HAL_ETH_TransmitIT {Frame}: failed! ret=%d err=%d",hal_ret,heth->ErrorCode);
> 		res = -EIO;
> 		goto error;
> 	}
> 
> 	if(k_sem_take(&dev_data->tx_int_sem, K_MSEC(ETH_DMA_TRANSMIT_TIMEOUT))!=0)
> 	{
> 		LOG_ERR("HAL_ETH_TransmitIT timeout on tx_int_sem take");
> 		res = -EIO;
> 		goto error;
> 	}
> 
> #else
> 
> 	hal_ret = HAL_ETH_TransmitFrame(heth, total_len);
> #endif /* CONFIG_SOC_SERIES_STM32H7X */
> 
> 	if (hal_ret != HAL_OK) {
> 		LOG_ERR("HAL_ETH_Transmit{Frame}: failed! ret=%d err=%d",hal_ret,heth->ErrorCode);
> 		res = -EIO;
> 		goto error;
> 	}
> 
> #if !defined(CONFIG_SOC_SERIES_STM32H7X)
> 	/* When Transmit Underflow flag is set, clear it and issue a
> 	 * Transmit Poll Demand to resume transmission.
> 	 */
> 	if ((heth->Instance->DMASR & ETH_DMASR_TUS) != (uint32_t)RESET) {
> 		/* Clear TUS ETHERNET DMA flag */
> 		heth->Instance->DMASR = ETH_DMASR_TUS;
> 		/* Resume DMA transmission*/
> 		heth->Instance->DMATPDR = 0;
> 		res = -EIO;
> 		goto error;
> 	}
> #endif /* CONFIG_SOC_SERIES_STM32H7X */
> 
> 	res = 0;
> 	error:
> 	k_mutex_unlock(&dev_data->tx_mutex);
> 
> 	return res;
> }
> ```
> 
> With these changes I have made the ethernet communication a lot faster and more reliable:
> 1. 0 packet losses for ICMP/ping with <2ms response time even with stress test 0.01s ping interval
>    1. ping -i 0.01 192.0.2.1 
>    2. 1038 packets transmitted
>    3. 1038 received, 0% packet loss, time 10373ms
>    4. rtt min/avg/max/mdev = 1.214/2.033/7.364/0.644 ms
> 
> 2. net telnet sample works flawlessly with no connection break
> 3. net socket coap-server sample works perfectly
> 4. net socket dumb-http-server works perfectly
> 
> 
> I hope you can integrate these into the driver meanwhile I learn how to use git and contribute correctly to the issues.

Very good work! I will test it later.
Of course the interrupt implementation should be way faster than polling rx. 
If you like I can give you short introduction into the working with git and zephyr repository management. With your good work - you earned it really. ;) Then I would suggest we create your own PR s.t. you can also learn more from my class and your code will stay under your name in the repo.
Can you imagine implement same approach for F4/F7 boards later?

erwango commented:
> Then I would suggest we create your own PR s.t. you can also learn more from my class and your code will stay under your name in the repo.

I completely agree

lochej commented:
> Very good work! I will test it later and give you feedback.
> Of course the interrupt implementation should be way faster than polling rx.
> If you like I can give you short introduction into the working with git and zephyr repository management over Skype or so. With your good work - you earned it really. ;) Then I would suggest we create your own PR s.t. you can also learn more from my class and your code will stay under your name in the repo.
> Can you imagine implement same approach for F4/F7 boards later?

Absolutely I would be very glad if you could teach me ! Then I'll contribute correctly using the good practices :) 
My approach with the HAL should be no different for the F4/F7 boards. 
Maybe a few HAL calls adaptation should do the trick.

Nukersson commented:
> > Very good work! I will test it later and give you feedback.
> > Of course the interrupt implementation should be way faster than polling rx.
> > If you like I can give you short introduction into the working with git and zephyr repository management over Skype or so. With your good work - you earned it really. ;) Then I would suggest we create your own PR s.t. you can also learn more from my class and your code will stay under your name in the repo.
> > Can you imagine implement same approach for F4/F7 boards later?
> 
> Absolutely I would be very glad if you could teach me ! Then I'll contribute correctly using the good practices :) 
> My approach with the HAL should be no different for the F4/F7 boards. 
> Maybe a few HAL calls adaptation should do the trick.

I will contact you later.

ABOSTM commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -38,6 +40,7 @@ config ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER
 	bool "Use DTCM for DMA buffers"
 	default y
 	depends on SOC_SERIES_STM32F7X
+	depends on !SOC_SERIES_STM32H7X
```
Due to F7 dependency, it is useless to introduce !H7
```suggestion
```

Nukersson replied:
The line were removed.

ABOSTM reviewed (changes requested):
1 remaining useless 'depends on'
otherwise, LGTM

Nukersson commented:
From my side - this PR can be merged.

ABOSTM reviewed (approved):
LGTM

Nukersson commented:
@lochej @erwango @ABOSTM 

Dear all I've tested the interrupt approach from @lochej on my `nucleo-h745zi-q` board. The branch itself is in [my repo](https://github.com/Nukersson/zephyr/tree/improved_stm32h7_eth_drv_supp) and is ready for @lochej to be integrated after the current PR is merged.

The performance were tested with following `ping` configuration on `1k ping packets` with `0.01 sec. ping interval`:
`ping -i 0.01 192.168.178.12 -c 1000 > log_file.log`

The board has static IP address and were connected to my `FritzBox 7590` router. My Mac laptop were connected over `5GHz WiFi` to the same router. The `distance` between the router and the laptop was just about `1 meter`. 

Note: Due to used WiFi my performance tests may vary from @lochej ones and can face lost packets.

***The results of the tests are the following:***

1. Old approach: Polling TX:
`1000 packets transmitted, 133 packets received, +2 duplicates, 86.7% packet loss
round-trip min/avg/max/stddev = 1.037/347.861/1002.657/372.646 ms`

2. New approach (thnx to @lochej ): Interrupt TX:
`--- 192.168.178.12 ping statistics ---
1000 packets transmitted, 508 packets received, 49.2% packet loss
round-trip min/avg/max/stddev = 1.154/65.999/742.486/150.473 ms`

Thus we see the increase in performance of factor ~2 (372ms / 150ms and 86.7%/49.2%) in **ping response deviation time** and **number of the lost packets**. Especially interesting is the **average ping response time** decrease by almost factor of ~5.3 = avg(348ms) / avg(66 ms). Thus the gain of the improvement can be clearly seen and the improvement from @lochej must be implemented.

P.S.: @lochej Please feel free to use [my repo](https://github.com/Nukersson/zephyr/tree/improved_stm32h7_eth_drv_supp) to create your own PR, since this my branch is only copy/paste of your solution.

Nukersson commented:
@erwango  Can you please resolve merge conflict? There is one very trivial conflict. Since I do not have access rights, I could not do it.

erwango commented:
@Nukersson, I can't do it neither, you have to fetch rebase and resolve the conflict.

Nukersson commented:
> @Nukersson, I can't do it neither, you have to fetch rebase and resolve the conflict.

Done. Some changes came just exactly after my last rebase today. :)

Nukersson commented:
@erwango Can I provide some support to increase merging?

pfalcon commented:
> Can I provide some support to speed up merging?

@Nukersson: As long as a PR collected 2 approvals, you can ping folks with merge rights, like @carlescufi or @nashif .

carlescufi commented:
@jukkar @gmarull any objections to merging this?

gmarull commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -94,3 +96,10 @@ config ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS
 
 endif # ETH_STM32_HAL
+
+config NOCACHE_MEMORY
```
Maybe this Kconfig setting should be *namespaced*?

Nukersson replied:
Can you give me some example for what you mean?

gmarull replied:
Some sort of Kconfig option prefix to avoid possible future name clashes (e.g. `ETH_STM32_NOCACHE_MEMORY`). I'm not sure if there is an established policy. @erwango @carlescufi ?

Nukersson replied:
Good question. I would wait until any suggestion from both comes.

erwango replied:
`ETH_STM32_NOCACHE_MEMORY` would do, thanks @gmarull for the catch.

Nukersson replied:
Changed as suggested.

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,19 +33,67 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#define LL_AHB1_GRP1_PERIPH_ETHMACRX	LL_AHB1_GRP1_PERIPH_ETH1RX
+#define LL_AHB1_GRP1_PERIPH_ETHMACPTP	LL_AHB1_GRP1_PERIPH_ETH1MAC
+
+#define ETH_DMA_TRANSMIT_TIMEOUT	20U
```
This timeout should be documented, maybe make it configurable.

Nukersson replied:
KConfig?

Nukersson replied:
I will put some description on this value, but suggest not to use it in KConfig. The main reason is interrupt based improvement form @lochej. This timeout value will be then eliminated. Are you okay with it?
P.S.: The suggested solution is already tested and prepared. @lochej and I just waiting until it this PR will be merged.

gmarull replied:
It's ok then, but still a small description would be good.

Nukersson replied:
Corresponding description were added.

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,19 +32,44 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#endif
+
+#ifndef PHY_BSR
+#define PHY_BSR		((uint16_t)0x0001U)
```
Constant should be documented/referenced

Nukersson replied:
The value of `PHY_BSR` comes from original implementation and without description/reference. Therefore I am not sure, I can describe it correctly.

Nukersson replied:
The corresponding define is copy-pasted from `stm32f7xx_hal_conf.h` as follows:
`#define PHY_LINKED_STATUS  ((uint16_t)0x0004U)  /*!< Valid link established */`

It defined only for `H7` series for sake of simplicity.

gmarull commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -441,7 +480,10 @@ static int eth_initialize(struct device *dev)
-	HAL_ETH_Start(heth);
+	hal_ret = HAL_ETH_Start(heth);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_Start{_IT} failed");
```
After this error function continues, is this wanted?

Nukersson replied:
Yes. My implementation just extends the original one by error logging.

gmarull reviewed (commented):
Added some comments, should not block PR though.

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,19 +32,44 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#endif
+
+#ifndef PHY_LINKED_STATUS
+#define PHY_LINKED_STATUS	((uint16_t)0x0004U)
```
The value of `PHY_LINKED_STATUS ` comes from original implementation and without description/reference. Therefore I am not sure, I can describe it correctly.

Nukersson replied:
The corresponding define is copy-pasted from `stm32f7xx_hal_conf.h` as follows:
`#define PHY_LINKED_STATUS  ((uint16_t)0x0004U)  /*!< Valid link established */`

Nukersson commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,19 +33,67 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #error DTCM for DMA buffer is activated but zephyr,dtcm is not present in dts
 #endif
 
+#ifndef PHY_ADDR
```
A separate layer to PHY SoC would be more clear way to implement ETH_PHY layer. This value actually belongs there. I will just delete `#ifndef`. Thus it will follow the original implementation in current master.

Nukersson replied:
Corresponding `#ifndef` will be removed within next push.

Nukersson replied:
There is no other place. Removed corresponding `#ifndef`.

Nukersson commented:
@gmarull @erwango Please approve the change in `KConfig.stm32_hal`. And please change approve state if everything is okay or let me know, if something need to be changed further.

gmarull commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -94,3 +96,11 @@ config ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS
 
 endif # ETH_STM32_HAL
+
+config ETH_STM32_NOCACHE_MEMORY
```
I think this setting should actually be removed. You can just add `select NOCACHE_MEMORY if SOC_SERIES_STM32H7X` in the `ETH_STM32_HAL_ETH`. I initially thought it was a setting specific to this module, but it's an actual Zephyr setting.

Nukersson replied:
Changed as suggested.

erwango commented:
@Nukersson can you check the CI failure ?

erwango commented:
Tagging DNM to avoid accidental merge until CI is fixed

Nukersson commented:
> @Nukersson can you check the CI failure ?
> 

Will be fixed today evening. Can't do it until then.

Nukersson commented:
> @Nukersson can you check the CI failure ?
> 

Will be fixed today evening. Can't do it until then.

Nukersson commented:
> @Nukersson can you check the CI failure ?

Fixed test. It runs locally. Let's see in few hours, what build kite shows.

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -1,14 +1,17 @@
 	bool "STM32 HAL Ethernet driver"
 	select USE_STM32_HAL_ETH
+	select USE_STM32_HAL_RCC_EX if SOC_SERIES_STM32H7X
+	select NOCACHE_MEMORY if SOC_SERIES_STM32H7X
```
NOCACHE_MEMORY depends (indirectly) on CPU_CORTEX_M7 arch, which restricts this driver implementation to M7 core on Dual Core H7 SoCs.
I'm fine with this, but this should be documented and verified in the code.

Nukersson replied:
Thanks for the info. You're right, `NOCACHE_MEMORY` makes sense only for M7 core, since it has DTCM coupled to it and M4 not. Any suggestions how `NOCACHE_MEMORY` can be coupled on M7 (e.g. select `NOCACHE_MEMORY if SOC_SERIES_STM32H7X && CPU_M7`)?

erwango replied:
`select NOCACHE_MEMORY if SOC_SERIES_STM32H7X && CPU_CORTEX_M7` should work.

Then, do you know if the driver will work on M4 core?
If not, it would be better to prohibit its use on M4 + H7 for now.

Nukersson replied:
I am pretty sure the driver will work on M4, but it is currently not tested. 
I would change then the line 11 as follows: `select NOCACHE_MEMORY if SOC_SERIES_STM32H7X && CPU_M7`
I will test now if `CPU_M7` Kconfig define or something similar is allowed. Will let you know within next new push.

Nukersson replied:
The line 11 changed to: `select NOCACHE_MEMORY if SOC_SERIES_STM32H7X && CPU_CORTEX_M7`

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -32,19 +33,60 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#define ETH_DMA_MEM	__aligned(4)
 #endif /* CONFIG_ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER */
 
+#if defined(CONFIG_NOCACHE_MEMORY)
+#define NO_CACHE __nocache
+#else
+#define NO_CACHE
+#endif  /* CONFIG_NOCACHE_MEMORY */
+
+static ETH_DMADescTypeDef dma_rx_desc_tab[ETH_RXBUFNB] NO_CACHE ETH_DMA_MEM;
+static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] NO_CACHE ETH_DMA_MEM;
+static uint8_t dma_rx_buffer[ETH_RXBUFNB][ETH_RX_BUF_SIZE] NO_CACHE ETH_DMA_MEM;
+static uint8_t dma_tx_buffer[ETH_TXBUFNB][ETH_TX_BUF_SIZE] NO_CACHE ETH_DMA_MEM;
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+static ETH_TxPacketConfig tx_config NO_CACHE;
+#endif
```
Just a comment on this one:
I wonder if there could be an alternative to `NO_CACHE` so that one doesn't conclude to fast that these symbols are not cached in any case

Nukersson replied:
1. According to [CONFIG_NOCACHE_MEMORY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_NOCACHE_MEMORY.html) `NOCACHE_MEMORY` just allocates non-cached memory area.

2. Taking 1 into account the corresponding memory buffers will be placed in this `NOCACHE_MEMORY` if it exist other way not. Thus exactly `nocache` functionality will be used here and `NO_CACHE` is okay I think.

3. I would just add following comment on `NO_CAHE` define: 
`/*  On some STM32 (e.g. H7 series) the non-cached memory needed to avoid ETH DMA data inconsistency. The NO_CACHE lets linker place corresponding buffers into non-cached memory area which must be previously created by CONFIG_NOCACHE_MEMORY KConfig definition */`

Are you okay with it?

erwango replied:
Im fine with the functionality, my comment was purely on the name of this macro, as `NO_CACHE` could actually mean 'cached' in some cases. I'd like to avoid potential confusions if possible

Nukersson replied:
You're right about naming. Especially on the line 78 it can be clearly seen: `NO_CACHE` used, but it is still cached.
**How about `CACHE` name of this define?** Then it does not says about certain usage of the memory (cached/non-cached) but still means memory type which can be some kind of cache her in context.

erwango replied:
Go for `CACHE`

Nukersson replied:
Done

erwango reviewed (changes requested):
Last mile..

lochej commented:
Congrats @Nukersson for this PR, the next step is for us to open another PR to add my interrupt based method to the driver 😄

Nukersson commented:
> Congrats @Nukersson for this PR, the next step is for us to open another PR to add my interrupt based method to the driver 😄

I will prepare to and contact you later.
